Add a reference guide for input program format

bader opened the pull request.

jinge90 committed:
Support llvm.is.constant

Signed-off-by: Ge Jin <ge.jin@intel.com>
Co-authored-by: Alexey Sachkov <alexey.sachkov@intel.com>
Co-authored-by: Artem Gindinson <artem.gindinson@intel.com>

Commit:
Initial implementation of new LLVM PassManager Passes.

This commit just introduces the new PassManager to the LLVMToSPIRV Translator.
Moved all the Pass functionality into base class and added two new derived classes, one for the LegacyPM one for the new PM.
Migrating to usage of the new PM will come in a subsequent commmit.

bader committed:
Add a reference guide for input program format

AlexeySachkov commented on docs/InputReferenceGuide.rst:
```diff
@@ -0,0 +1,157 @@
+|                    |    ; 6 - Generator Id, 123 - Generator Version          |
++--------------------+---------------------------------------------------------+
+
+**TODO**: Should we document all OpExecutionMode operands supported by the
```
Moreover, even ones documented above are not required and all of them are optional, i.e. used to override the default behavior of the translator

AlexeySachkov commented on docs/InputReferenceGuide.rst:
```diff
@@ -0,0 +1,157 @@
+are naturally represented as LLVM IR address spaces with following mapping:
+
+SPIR-V storage class    LLVM IR address space
+Function                No address space
```
```suggestion
Function                No address space or addrspace(0)
```

AlexeySachkov commented on docs/InputReferenceGuide.rst:
```diff
@@ -0,0 +1,157 @@
+Unsupported LLVM IR features
+============================
+
+**TODO**: Should these features be listed explicitly?
```
It might be easier to list supported features, because it is smaller and it is not updated so often as list of unsupported ones (LLVM development moves faster than the translator development)

AlexeySachkov commented on docs/InputReferenceGuide.rst:
```diff
@@ -0,0 +1,157 @@
+which defines most of the rules for encoding SPIR-V instructions and types in
+LLVM IR.
+
+**TODO**: should we merge these two documents?
```
I think it might make sense, because "SPIR-V representation in LLVM IR" documents mapping from LLVM IR to various SPIR-V features. For example, if you want to get a barrier or memory fence in SPIR-V, you need to read that document to see which LLVM IR should be generated

bader committed:
Apply suggestions from code review

Co-authored-by: Alexey Sachkov <alexey.sachkov@intel.com>

Nuullll committed:
Expose SPIRVToOCL12, SPIRVToOCL20 interfaces in SPIRVToOCL.h

Signed-off-by: Yilong Guo <yilong.guo@intel.com>

mlychkov committed:
Add support for constant integers of any bitwidth

Add translation of constant integers of any bitwidth greater or equal 2.
However, they are still limited to maximum supported number of words in
one SPIR-V instruction. For constants maximum word count for literal
part is 65532. So, max integer bitwidth is 65532*32.

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

bader committed:
Applied code review suggestions from Alexey

Merged "Input reference guide" into SPIRVRepresentationInLLVM document.

Fznamznon committed:
Fix LLVM-SPIRV::dbg-value-location.ll test

After recent changes in LLVM, debug info for this test changed for
Windows triple and caused test fail on Windows. Debug info can be
represented in several ways, it may depend on the triple. This test was
originally made for x86_64-apple-darwin10.0.0 triple, so this patch sets
this triple permanently.

dm-vodopyanov committed:
Fix test/DebugInfo/X86/static_member_array.ll

This patch fixes static_member_array.ll test which fails on Windows. The
reason why it failed is missing a-f letters in hex number from regex in
test.

zahiraam committed:
Fixed various compiler errors.

Our compiler is triggering various errors. This is a patch to fix them.

AlexeySotkin committed:
Fix translation of cmpxchg instruction

Previous implementation optimized produced code assuming a particular
pattern in input LLVM IR. But that assumption is not always correct.
Now we always insert extra instructions to preserve semantics of cmpxchg.
Those extra instruction can be optimized later.

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

Fznamznon committed:
Implement SPV_INTEL_optnone extension (#913)

This extension adds new Function Control mask value to request to not
optimize the corresponding function. It supposed to work as analog of
`optnone` function attribute in LLVM IR.

The spec can be found under https://github.com/intel/llvm/pull/3198

Commit:
Ignore experimental_noalias_scope_decl intrinsic (#906)


Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Support the SPV_EXT_shader_atomic_float_min_max extension (#916)

* Support the SPV_EXT_shader_atomic_float_min_max extension

As with other atomic instructions, we're looking to translate
an LLVM IR call akin to `__spirv_AtomicFMaxEXT()` into a
corresponding SPIR-V OpCode akin to `AtomicFMaxEXT`.

Extension specification can be found at
https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/EXT/SPV_EXT_shader_atomic_float_min_max.asciidoc

Support in SPIR-V Header was added with
https://github.com/KhronosGroup/SPIRV-Headers/pull/189.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

Fznamznon committed:
Fix SPIR-V friendly LLVM IR for conversion functions (#899)

In most cases convert functions are translated to set of LLVM IR
instructions but in case of saturated conversion or if conversion has
rounding mode, the translation should go through SPIR-V friendly LLVM IR.

bader commented:
@AlexeySachkov, I'm going to open a new PR for Khronos repository. Any objections?

AlexeySachkov commented:
> @AlexeySachkov, I'm going to open a new PR for Khronos repository. Any objections?

@bader, I don't have any objections

Commit:
Accept intrinsic namings into -spirv-allow-unknown-intrinsics (#912)

Previously, enabling the option would let all intrinsics through
the translator. From now on, the option can be configured so
that only specific intrinsic calls with the given prefix are
allowed, while those not matching the prefix will result in an
error as if no option was given.

The allowed intrinsic prefixes are fed into the option as a
comma-separated list, akin to `--spirv-ext`. The list can also
be left unspecified, which, as before, will accept all intrinsic
calls into SPIR-V.

One of the relevant use cases involves llvm.genx.* intrinsics,
which stem from SYCL ESIMD code, but can also mix with common LLVM
intrinsics that stem from regular SYCL code. It would be desirable
to allow ESIMD-specific intrinsics, but still error out on other
unknown ones, To better highlight this case, all ESIMD-specific
tests have been switched to passing an explicit prefix to the
CLI option.

AlexeySachkov committed:
Disable Travis CI on master branch

We have GitHub Actions configured and there is no need to use two CI
systems at the same time.

The .travis.yml config file is not removed completed yet as we haven't
swithed to use GitHub Actions on release branches - this file will be
kept as some kind of reference for now.

AlexeySachkov committed:
[GitHub Actions] Fix paths-ignore

Changed `xml` -> `yml` to avoid launching some workflows on unrelated
changes, like we don't need to launch out-of-tree build if the only
changed file in the PR is in-tree build configuration.

AlexeySachkov committed:
Remove Travis CI badge

Commit:
Minimize IR for atomic float tests (#917)

Removes excessive kernel code from the test examples. The key
source-level change for re-generating such IR from Intel SYCL is
to use `SYCL_EXTERNAL` function definitions instead of "true"
heterogeneous code. E.g. the following DPC++ code may be used to
generate IR for the AtomicFAdd test:

```
using namespace cl::sycl;

template <typename T>
using atomic_ref_T = ONEAPI::atomic_ref<T, ONEAPI::memory_order::relaxed,
                     ONEAPI::memory_scope::device,
                     access::address_space::global_space>;

SYCL_EXTERNAL decltype(auto) AtomicFloatInc(float &Arg) {
  atomic_ref_T<float> Atm(Arg);
  return Atm.fetch_add(1);
}

SYCL_EXTERNAL decltype(auto) AtomicDoubleInc(double &Arg) {
  atomic_ref_T<double> Atm(Arg);
  return Atm.fetch_add(1);
}

gfxstrand committed:
Implement read/write_mem_fence

There's no real difference at the SPIR-V level between these and the
more generic mem_fence intrinsic because mem_fence is already relaxed
and SPIR-V doesn't have a concept of anything more relaxed than that.

Closes: #920

svenvh committed:
Fix -Wignored-qualifiers warnings

Fix two instances of:

  warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]

mlychkov committed:
[SYCL][FPGA] Update ArbitraryFloatCastTo/FromInt processing

Add processing of a new parameter To/FromSign that specifies if integer value
to cast to/from is signed or not for ArbitraryFloatCastTo/FromInt functions.

See spec update (intel/llvm#1934): https://github.com/intel/llvm/blob/ea96f3ccee007e04074af6e275dbfce4fce13717/sycl/doc/extensions/SPIRV/SPV_INTEL_arbitrary_precision_floating_point.asciidoc

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

svenvh committed:
Update DebugInfo test after LLVM change

Update for LLVM commit d39bc36b1be7 ("[debug-info] refactor
emitDwarfUnitLength", 2021-02-25).

svenvh committed:
Use -fdeclare-opencl-builtins for more .cl tests

Since clang commit 85eb12eefdf6 ("[OpenCL] Add declarations with
enum/typedef args", 2021-02-24), Clang provides most of the OpenCL
builtin functions through the faster `-fdeclare-opencl-builtins` flag.
Update various .cl tests to make use of this flag, to speed up running
of the tests and to avoid maintaining excerpts from `opencl-c.h` in
the tests.

Nuullll committed:
[DbgTran] Fix wrong usage of replaceTemporary

The uniqued node could have a different address.

Signed-off-by: Yilong Guo <yilong.guo@intel.com>

KornevNikita committed:
[SPIRV] Encode debug info producer in SPIR-V (#924)

* [SPIRV] Encode debug info producer in SPIR-V

This is a workaround to keep debug info producer information throughout the
translation. We need it because the debug info specification doesn't provide
means to carry this information.

saininav committed:
cmake: allow to enable/disable ccache

By default ccache is enabled.

Signed-off-by: Naveen Saini <naveen.kumar.saini@intel.com>

MochalovaAn committed:
Remove language check assert

According to specification, OpSource has no semantic impact and can safely be removed from a module. Therefore, such a strict check is not required.

Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

AlexeySotkin committed:
Redesign handling of forward pointers

In some cases the current implementation doesn't create forward pointers
when they are required. New approach is to create forward pointers
during sorting of the type instructions.

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

Commit:
Fix builds: adjust to the new DbgVariableIntrinsic API

The API change is brought by llvm/llvm-project@e5d958c.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

MochalovaAn committed:
Add Early Exit

Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

NikitaRudenkoIntel committed:
Add HandleLLVMOptions to main CMakeLists.txt
to propagate configuration flags. In particular, this enables
Multi-threaded option in MSVC (instead of Multi-threaded DLL)

MochalovaAn committed:
Fix llvm-spirv crash when count of Fortran metadata variables is an array

Handle the case when Count of Fortran metadata variables is an array

Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

KornevNikita committed:
[SPIRV] Encode CodeView checksum information in SPIR-V (#936)

* [SPIRV] Encode CodeView checksum information in SPIR-V

This is a workaround to keep CodeView checksum information throughout the
translation. The idea is to store checksum kind and value to Ops[TextIdx]
and then parse this string while creating DIFile.

aparshin-intel committed:
Try to derive the size of vectors when translating DICompositeType

SPIRV translator does not have the facilities to preserve the size specified
in the DICompositeType node for DW_TAG_array_type.  Instead, it relies on the
size of the underlying base type to calculate the intended size. In
cases when the underlying base type is typedef - this size can`t be
restored. This commit introduces a way to recursively traverse the
definition of the base type to restore the original size

NikitaRudenkoIntel committed:
Fix uninitialized variables warnings
These warnings come into place when HandleLLVMOptions is imported

aratajew committed:
Fix mangling of OpAtomicUMax/OpAtomicUMin representation in SPV-IR

Commit:
Print path info in error messages only in debug build (#941)

* Print path info in error messages only in debug build

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

aratajew committed:
Fix emission of SPIR-V friendly IR for vload/vload_half/vload_halfn/vloada_halfn OpenCL EIS instructions

Fznamznon committed:
Add support for SPV_KHR_linkonce_odr (#950)

This extension adds a new Linkage Type LinkOnceODR. A global variable or
function with such linkage type can be merged with equivalent global variables
or functions from other modules during the linking process. The primary use case
of this linkage type is to satisfy the C++ language One Definition Rule.

The spec
https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_linkonce_odr.asciidoc

Commit:
[NFC] Remove unnecessary addCapability calls for decorations

It's already handled in SPIRVEnum.h.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Fznamznon committed:
Fix SPIR-V friendly IR for OpGroupAsyncCopy and OpGroupWaitEvents (#963)

These instructions were translated directly to OCL even if SPIR-V
friendly LLVM IR was requested.

MochalovaAn committed:
Translate the llvm.fshr intrinsic function

Add llvm.fshr intrinsic translation, similar to llvm.fshr translation.
Add test for that
A detailed description of FSHR can be found at
https://llvm.org/docs/LangRef.html#llvm-fshr-intrinsic

Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

MochalovaAn committed:
Add llvm.*max and llvm.*min intrinsics translation (#973)


Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

bashbaug committed:
fix legacy atomic and memory fence memory order and scope (#965)

* fix memory scope and memory order for legacy atomics
* fix memory order for legacy memory fence functions

Co-authored-by: Mikhail Lychkov <51128024+mlychkov@users.noreply.github.com>
Signed-off-by: Ben Ashbaugh <ben.ashbaugh@intel.com>

Commit:
Remove mapping of NoWrite function parameter attribute on qual metadata

The NoWrite SPIR-V function parameter provides an assurance that
memory pointed to by a NoWrite pointer will not be modified through
that pointer.

The const qualifier (represented in LLVM IR for kernel functions
using kernel_arg_type_qual metadata) does not provide any such
guarantee. There is nothing that prevents a function that takes a
pointer to a const pointee from stripping the const pointer away
and modifying the pointee.

As such, it is not appropriate to translate from one to the other,
and so the code for doing so should be removed, so that this
translation is not performed either when translating from LLVM IR
to SPIR-V, or when translating from SPIR-V to LLVM IR.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Map NoWrite function parameter attribute on LLVM readonly

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Simplify the test

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Promote SPV_INTEL_optimization_hints to SPV_KHR_expect_assume

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix macro

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Implement SPV_INTEL_memory_access_aliasing extension (#966)

* Implement SPV_INTEL_memory_access_aliasing extension

This extension adds analogs of alias.scope and noalias metadata
in SPIR-V. These metadata is being used to annotate non-aliasing
memory accesses and being generated, for example, during inlining
of functions with restrict pointer parameters.

Spec: https://github.com/intel/llvm/pull/3426

This patch also casually adds OpDecorateId in the translator flow.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Add missing lines

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

svenvh committed:
[CI] Run check-code-style on explicit OS version

Instead of specifying "latest", which GitHub may change any time,
specify the OS version that the checks actually depend on.

Fznamznon committed:
Fix SPIRV friendly IR for OpImageSampleExplicitLod&OpSampledImage (#984)

These instructions were translated directly to OCL even if SPIR-V
friendly LLVM IR was requested.

Commit:
Don't remove SPV_INTEL_optimization_hints declaration

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

KornevNikita committed:
Fix crash at translation of Entity of DebugImportedEntity (#951)

This is workaround to fix crash, which caused by using `dyn_cast` on a `nullptr`,
as the result of translation of `Entity` operand of `DebugImportedEntity`, because
this operand is `DebugInfoNone`.

svenvh committed:
Update for LLVM iterator change

Fix the build after LLVM commit 0a92aff721f4 ("Replace uses of
std::iterator with explicit using", 2021-04-12).

Commit:
Tolerate more inputs during alias.scope/noalias MD translation (#994)

According to https://llvm.org/docs/LangRef.html aliasing metadata
must have following layout:
List { MD Node Scope1, (other scopes) ... }
Scope1 { MD Node Scope1, MD Node Domain1, (optional MD String) }
... (other scopes)
Domain1 { MD Node Domain1, (optional MD String) }
... (other domains)

and this pattern is actually being checked in LLVM's
ScopedNoAliasAA.cpp.

But in a harsh reality LLVM opt have bugs, which can result in this
pattern violation. So lets be more tolerant to the input IR module.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix incorrect translation of FPGA decoration on arrays (#983)

* Fix incorrect translation of FPGA decoration on arrays

var.annotation shall be emitted, but instead the translator was
generating ptr.annotation. It was happening due to incorrect assumption,
that for attributed static memory variable decoration shall be applied
to alloca instruction, whilst in the reality the alloca instruction
might be hidden by several cast instructions.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

KornevNikita committed:
Fix transTypeComposite bug (#964)

* Fix transTypeComposite bug

This is a workaround to fix an error caused by DIBuilder which always creates
a node with the DW_TAG_struct_type tag for both class and structure.

Commit:
Add W/A to translate OpenCL kernel_arg_type_qual metadata

Since constant qualifier stored in the metadata is no longer mapped
on NoWrite function parameter attribute we need a another way to
save information about it. As work around this patch stores argument
qualifier information in OpString under
'preserve-ocl-kernel-arg-type-metadata-through-string' option. This
option should be also passed during reversed translation to obtain
the metadata.

This patch also moves already existing workaround for kernel_arg_type
metadata under the option.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix clang-format

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Nuullll committed:
[DbgTran] Translate subprogram debug info before scopes

This resolves the parent scope dependency issue of lexical blocks.

Signed-off-by: Yilong Guo <yilong.guo@intel.com>

svenvh committed:
[CI] Exclude spirv.hpp for clang-format

Commit:
Add 'willreturn' attribute to OpenCL work-item functions

These functions will definitely return result, so lets annotate
them with the attribute during reverse translation.

Also, in https://reviews.llvm.org/D96993 behavior of DCE pass was
changed. Now it doesn't remove non-willreturn calls. So for example
in a case when user called get_global_id(dimidx) several times -
redundant calls without 'willreturn' attribute won't be eliminated.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix warning when build by clang (#1004)

Warning message:
private field 'RetTy' is not used [-Werror,-Wunused-private-field]

Signed-off-by: Pavel V Chupin <pavel.v.chupin@intel.com>

aratajew committed:
Move OCL-IR translation logic related to EIS instructions to SPIRVToOCL pass

This change also fixes vstore translation for SPV-IR generation path and adds
test for vstore and printf.

Commit:
[NFC] Fix FPGALoopAttr.ll after fd94103

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

AlexeySotkin committed:
Replace type pointers by ids in SPIRVTypeFunction

The type of a function parameter may be id of a forward declared pointer type.
When we decode such module, we can't construct SPIRVTypeFunction, because
actual SPIRVTypePointer has not been decoded yet. Instead we can keep ids of
paramter types which can be resolved later.

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

wenju-he committed:
Fix -Werror=non-virtual-dtor -Werror=suggest-override errors (#1016)

Fixed the following warnings:
- "has virtual functions and accessible non-virtual destructor"
- runOnModule "can be marked override"

Fznamznon committed:
[NFC] Remove dead code

This code is not needed after 9f3c10f4.

Fznamznon committed:
Fix SPIR-V friendly IR for event opcodes

This instruction was translated directly to OCL even if SPIR-V
friendly LLVM IR was requested.

MochalovaAn committed:
Fix memory leak after getAsInstruction

Adding a destructor and a special class field to account for getAsInstruction instruction.
Per LLVM documentation, getAsInstruction doesn't attach the result to any Module/Function/Basic Block and therefore it is not deleted as part of their destructors; basically, any pointers to this memory are lost at the end of corresponding if scopes.

Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

AlexeySachkov committed:
Relax handling of boolean specialization constants (#1015)

When SPIR-V friendly IR is used, it might be the case that it was
generated by some other LLVM tool and/or pass and for boolean types it
is possible that they are represented as i8 instead of i1 in LLVM IR
(depending on how they were used in the original source file).

This patch adjusts and relaxes handling of boolean specialization
constants from SPIR-V friendly IR to get the following signatures
acceptable: `i1 _Z20__spirv_SpecConstantia(i32, i8)`.

svenvh committed:
Add missing atomic include

std::memory_order used in this file is defined in <atomic>, which was
not included directly.

Nuullll committed:
Expose SPIRVToOCL12, SPIRVToOCL20 interfaces in SPIRVToOCL.h (#1023)

Expose SPIRVToOCL12, SPIRVToOCL20 interfaces in SPIRVToOCL.h

This "recommit"s #904, which was accidentally "revert"ed by #840

Signed-off-by: Yilong Guo <yilong.guo@intel.com>

svenvh committed:
Only fail getSpecConstInfo if badbit is set

Attempting to read past EOF may set both the `eofbit` and the
`failbit`, neither of which should be fatal in this case.  This could
cause test/OpSpecConstant.spvasm to fail spuriously.

svenvh committed:
Fix placeholder in test

%from-spirv is not a valid substitution and just results in a file
with that name.

svenvh committed:
Remove execute bit from test files

svenvh committed:
[NFC] Move top level SpecConstant tests into dedicated dir

svenvh committed:
Support OpSpecConstantOp in OpVariable initializer

The Initializer can be a constant instruction, which includes
OpSpecConstantOp instructions.

svenvh committed:
Support OpSpecConstantOp in array type length

Fznamznon committed:
Fix SPIR-V friendly IR for OpGenericCastToPtr instruction (#1019)

* Fix SPIR-V friendly IR for OpGenericCastToPtr instruction

This instruction was translated directly to OCL even if SPIR-V
friendly LLVM IR was requested.

svenvh committed:
Add OpSpecConstantOp test

Fznamznon committed:
Fix SPIRV-IR for OpImageWrite instruction (#1018)

* Fix SPIRV-IR for OpImageWrite instruction

Do not translate it directly to OCL.

Commit:
Implement SPV_INTEL_token_type extension

This extension adds a type-declaration instruction OpTypeTokenINTEL,
which is an analog of LLVM's type token. This type specifies IDs of
entry points and exit points of certain IL constructs, for example it
can be useful for exception handling representation in IL.

Spec: https://github.com/intel/llvm/pull/3788

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Lower atomic FP __spirv builtins into OpenCL 1.2 & 2.0 builtins

In backwards translation, we look to treat FP-typed atomic functions
from multi-vendor extensions similarly to "regular" atomics.

Co-authored-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>
Co-authored-by: Artem Gindinson <artem.gindinson@intel.com>

Commit:
Facilitate pre-installed Clang detection for in-tree builds (#1042)

Comply with LLORG community changes in llvm/llvm-project@a1e6565.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

Fznamznon committed:
Map OpVariable to LLVM value before initializer translation

When some global variable contains constant function pointer to a
function and this global variable is referenced from the pointed
function, translation crashed because intializer for global variable was
translated before mapping of OpVariable to LLVM value. It happened
because during translation of initializer in such case the original
variable was met for the second time, but it wasn't mapped to a proper
llvm value.

Commit:
Implement SPV_INTEL_fpga_invocation_pipelining_attributes extension

Some FPGA devices and toolchains can support customizable levels or
implementation of pipeline parallelism when mapping a SPIR-V module
to hardware. Through pipeline parallelism, multiple invocations of
a kernel or function can execute concurrently.

This extension adds decorations to request that a kernel or function
support invocations at a specified initiation interval, that multiple
invocations are forbidden from executing concurrently, or that the
kernel or function is limited to a maximum number of concurrent
invocations.

Spec:
https://github.com/KhronosGroup/SPIRV-Registry/blob/38a7e56b9b7feab7fa8b3bea08770f7232b76ed2/extensions/INTEL/SPV_INTEL_fpga_invocation_pipelining_attributes.asciidoc

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

KornevNikita committed:
Fix for ParseChecksum function (#1021)

* Fix for ParseChecksum function

This is a patch to fix failure, caused by bringing as argument of ParseChecksum
StringRef with non-zero length, but with empty data.

KornevNikita committed:
Add error generation when using memmove pass with PHINode as source (#1041)

* Report error when llvm.memmove is used with PHINode as the source operand

This is a patch to describe a failure in visitMemMoveInst function caused by
using llvm.memmove with PHINode as the source.

KornevNikita committed:
[SPIRV] New FPGA function attribute for DSP control (#1046)

This patch adds to SPIR-V translator support of the new FPGA function attributes
for DSP control of math operations.
Spec: KhronosGroup/SPIRV-Registry#100

Commit:
Fix clang-tidy concerns for FP atomic function lowering

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

Commit:
[SPIR-V] Stop re-writing SPIR-V type map (#1047)

* [SPIR-V] Stop re-writing SPIR-V type map

There is a map used during forward translation. It links LLVM type
and SPIR-V type, that is being generated, LLVM type here is used as
a key. Due to a bug in type mapping (information is below), it's
possible that for a single pointer type in LLVM, the translator
will create 2 physically-indentical SPIR-V pointer types and with
previous implementation of mapType function the map would be changed
during the second type insertion. This patch prevents it, forcing
to reuse the type that was created first, though we would still have
an unused duplication of this type in SPIR-V module.

Please consider following LLVM IR sample:
%struct._ZTS4Args.Args = type { %struct._ZTS6Object.Object addrspace(4)* }
%struct._ZTS6Object.Object = type { i32 (%struct._ZTS6Object.Object addrspace(4)*, i32)* }

Here we have 2 LLVM type declarations. The translator will recursively
go through structure members and then through element types of the
pointer types, creating the appropriate SPIR-V types and mapping them
on LLVM types. Due to a recursive nature of this algorithm, here we will
have '%struct._ZTS6Object.Object addrspace(4)*' be processed twice with
2 SPIR-V type declaration instruction being created.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

svenvh committed:
Handle OpSpecConstantOp for SNegate, Not, and LogicalNot

svenvh committed:
Use IRBuilder for folding

Rely on the IRBuilder to fold into constant expressions if possible.

svenvh committed:
Extend OpSpecConstantOp testing for initializers

Co-authored-by: Stuart Brady

svenvh committed:
Handle OpSpecConstantOp for integer comparisons

This revealed that OpINotEqual was missing from
isSpecConstantOpAllowedOp.

Test co-authored by Stuart Brady.

Commit:
Change return value of mapType function

It's a follow up change for https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1047

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

svenvh committed:
[NFC] Use SPIRVBinaryInst for OpFMod and OpSMod

OpFMod and OpSMod were implemented using a separate class, but
seemingly without an underlying motivation.  Instead of a separate
class, reuse the SPIRVBinaryInst class to unify OpFMod and OpSMod
handling with other binary instructions.

This change enables supporting FMod and SMod operators with
OpSpecConstantOp, which will be done in a future commit.

Commit:
Move SPV_INTEL_fast_composite tokens into internal:: namespace

This is being done as one of the migration steps to downloading
spirv.hpp from SPIR-V Registry during the build time.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

svenvh committed:
Add tests for SConvert, UConvert, BitCast OpSpecConstantOp

These do not need any further implementation work.

svenvh committed:
Handle OpSpecConstantOp with VectorShuffle

Modify the `SPIRVShuffleVector` constructor to allow passing a nullptr
basic block (as is the case for variable initializers).

Modify the `SPIRVShuffleVector` constructor to take `SPIRVId`s instead
of `SPIRVValue`s, which better reflects what is stored in the class,
and saves us an unnecessary ID-to-Value-to-ID round trip in
`createInstFromSpecConstantOp`.

svenvh committed:
Handle OpSpecConstantOp with CompositeExtract and CompositeInsert

Modify the constructors to allow passing a nullptr basic block (as is
the case for variable initializers).

Modify the constructors to take `SPIRVId`s instead of `SPIRVValue`s,
which better reflects what is stored in the class, and saves us an
unnecessary ID-to-Value-to-ID round trip in
`createInstFromSpecConstantOp`.

Modify test/opundef.spt to return the constructed value, so that it
does not get optimized out.

svenvh committed:
Update DebugInfo test for LLVM address space change

After LLVM commit 333987b04589 ("[OpenCL] Add DWARF address spaces
mapping for SPIR", 2021-06-04), the address space is now part of the
debug info.  Update the pattern to accommodate it.

AlexeySotkin committed:
[SPIR-V] Fix duplicate pointers

Current implementation creates forward pointer only if direct pointee
was already discovered during type traversal. But it is possible to have a
cyclic dependency when pointer points to already discovered type via multiple
intermediate types.

Modified algorithm tracks cyclic dependency case via return value of the
recursive algorithm.

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

KornevNikita committed:
Fix translating of compile unit (#1057)

* Fix translating of compile unit

It's a patch to fix ignoring of checksum info when translating compile unit

svenvh committed:
[CI] Align clang-tidy/format versions to LLVM version

Using a different LLVM version to obtain the compile commands may
cause inclusion of compiler flags that aren't recognized by
the older version of clang-tidy.

svenvh committed:
Handle OpSpecConstantOp with SMod

AlexeySotkin committed:
Fix attachment of decoration to spec constants

If a phi instruction uses result of __spirv_SpecConst() call which has
not been translated yet, we created a OpForward instruction. Then we
translated the call to OpSpecConstant and attached the SpecId decoration
to it. After that, when we map LLVM value to SPIR-V values we detected
that there is a OpForward instruction and replaced Id of OpSpecConstant
with Id of the OpForward instruction, leaving the decoration targeted to
non-existing instruction.

To avoid this incorrect behaviour we should attach decorations after
replacing OpForward instruction, i.e. in transDecoration method.

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

Fznamznon committed:
Implement support for dynamic memmove (#1060)

* Implement support for dynamic memmove

Use LLVM expandMemMoveAsLoop function in case memmove intrinsic use has
non-constant length.

svenvh committed:
Handle OpSpecConstantOp with Select

`OpSpecConstantOp` may be used outside of a basic block (for example
in a variable initializer).  Modify the `SPIRVSelect` constructor to
accept a `nullptr` basic block.

Use an IRBuilder when translating `OpSelect`, so that the resulting
select expression will be created as a constant expression (and
possibly constant-folded) where possible.

Update some tests as the SPIR-V to LLVM translation now folds selects
more aggressively due to the use of IRBuilder.

Fznamznon committed:
Fix handling of complex constant expressions (#1071)

When some constant expression is used as operand of regular instruction
and operand of other constant expression, the current algorithm of
lowering could produce incorrect order of instructions because it is not
possible to add instruction operand to a constant expression. Consider
following pseudo code example:
```
call foo(constexpr2(constexpr1), constexpr1)

// After first loop iteration through operands of the call instruction:

A = constexpr2op(constexpr1)
call foo(A, constexpr1)

// Ok, instruction A is now a user of constexpr1, so, when second
// operand is processed, all uses of constexpr1 are updated, but the
// instruction that represents constexpr1 is inserted before call
// instruction because it is now being processed, so it will look like
// this:

A = constexpr2(B)
B = constexpr1
call foo(A, B)

// So, instruction B needs to be moved after all its users to get a
// valid module:

B = constexpr1
A = constexpr2(B)
call foo(A, B)

```

svenvh committed:
[NFC] Remove switch on opcode by inheriting from SPIRVInstTemplateBase

Make the following classes inherit from SPIRVInstTemplateBase:
 - SPIRVSelect
 - SPIRVVectorShuffle
 - SPIRVCompositeExtract
 - SPIRVCompositeInsert

This allows removing a switch on opcode in SPIRVInstruction.cpp.

Fixes https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1072

svenvh committed:
Update for sys::fs::OpenFlags LLVM API change

Update for LLVM commit 3302af9d4c39 ("Support: Remove
F_{None,Text,Append} compatibility synonyms, NFC", 2021-06-15).

aratajew committed:
Fix SPIRV-IR for group opcodes

Commit:
Replace assumptions during llvm.loop translation with checks

There were 3 assumptions made during translation of llvm.loop metadata
into LoopMerge instruction:
1. A latch block for 'for' and 'while' loop ends with a unconditional
   branch instruction;
2. A latch block for 'do-while' loop ends with a conditional branch
   instruction;
3. For a 'conditional' latch basic block it assumes, that the first
   successor is the loop header and the second one is the exit block.

All three of them can be violated in a case of an optimized IR is passed
to the translator. In this case LoopMerge:
a. can be placed in a wrong basic block;
b. have continue target and merge block parameters sharing the same id.

This patch makes the translator to assume less and do more checks via
LLVM LoopInfo infrustructure.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

aratajew committed:
Remove unncessary LLVM_FALLTHROUGH statements for empty switch cases

svenvh committed:
Fix OpModuleProcessed memory leak

Commit:
Apply suggestion

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix comments

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Implemented SPIR-V support for loopcount attributes (#1062)

* Implemented SPIR-V support for loopcount attrs

There are 3 metadata, which map on LoopControlLoopCountINTELMask. Mask
has 3 parameters. Default values are -1.

Signed-off-by: Leonid Pauzin <leonid.pauzin@intel.com>

Commit:
Improve handling of memmove when Lenght > size of the type to move

It is possible that for a single element Src type size doesn't
match memory size to move (for example we copy N elements of one
integer array to an address of a first element of another integer
array).

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix SYCL/OpenCL vector size in DebugInfo reverse translation (#1082)

* Fix SYCL/OpenCL vector size in DebugInfo reverse translation

The SYCL/OpenCL specifications require that 3-element vectors
are sized equally to the 4-element ones (see 4.10.2.1 and 4.10.2.6,
khronos.org/registry/SYCL/specs/sycl-1.2.1.pdf). Meanwhile, the
logic for translating SPIR-V debug info into LLVM IR obeys a
trivial `vec_size = num_elems * elem_size` formula, which fails
for the 3-element edge case.

Example LLVM IR pre-translation:
```
!0 = !DICompositeType(tag: DW_TAG_array_type, baseType: !3, size: 128, flags: DIFlagVector, elements: !1)
!1 = !{!2}
!2 = !DISubrange(count: 3)
!3 = !DIBasicType(name: "int", size: 32, align: 32, encoding: DW_ATE_signed)
```
Faulty LLVM IR after bi-directional translation (note the size in `!0`):
```
!0 = !DICompositeType(tag: DW_TAG_array_type, baseType: !3, size: 96, flags: DIFlagVector, elements: !1)
!1 = !{!2}
!2 = !DISubrange(count: 3)
!3 = !DIBasicType(name: "int", size: 32, align: 32, encoding: DW_ATE_signed)
```
Until SPIR-V DI instructions are re-designed to store the array
size information, handle the 3-element case explicitly to favor
OpenCL/SYCL requirements.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

svenvh committed:
Emit spirv_internal.hpp include

svenvh committed:
[NFC] Sync spirv.hpp with SPIRV-Headers

Bring spirv.hpp in sync with f95c3b3 ("Merge pull request #219 from
cmarcelo/SPV_EXT_shader_atomic_float16_add", 2021-06-23) from
github.com/KhronosGroup/SPIRV-Headers .

Notably, this brings the SPV_KHR_integer_dot_product extension enum
values and CPP_for_OpenCL source language enum value, together with
other additions.  There are some reorderings too.

svenvh committed:
[NFC] Regenerate *Enum.h files after spirv.hpp sync

Regenerate SPIRVIsValidEnum.h and SPIRVNameMapEnum.h using
tools/spirv-tool/gen_spirv.bash after syncing spirv.hpp and manually
fix the following:

 - internal:: values are not handled, so they had to be added manually
   again.  Move all internal values to the end of the generated
   enum/function, so that they are together.

 - NameMap entries for IOPipesINTEL and FuncParamIOKindINTEL do not
   follow the convention of the other values, so they had to be
   changed back to their original values.

svenvh committed:
Change SPIRVInstTemplate creation order

In a followup commit, an instruction's required capabilities will
depend on its argument types.  This does not work with the current
LLVM-to-SPIR-V translation that consists of the following steps:
 1. Create the instruction without operands.
 2. Add it to the module (which queries the instruction's required
    capabilities).
 3. Add operands.

Swap steps 2 and 3, such that the instruction is fully constructed
before querying its required capabilities.

Commit:
Remove alias.scope/noalias exclusive restriction (#1061)

It should be a valid case if alias.scope and noalias mask/decoration
are applied to the same instruction.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Fix TopologicalSort memory leak (#1092)

Commit:
[RFC] Improve code for IVDep backwards translation (#1084)

* [RFC] Improve code for IVDep backwards translation

This refactoring phase comes down to moving the translation algorithm
out of the already-cluttered `SPIRVToLLVM::setLLVMLoopMetadata()` body.

For now, a static member-only class is employed: it provides encapsulation
for helper functions while avoiding the unnecessary complexity that "true"
class entities would bring. A simple `LoopsEmitted` set is used to guard
against duplication of metadata for a particular loop by the callers.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

Commit:
[NFC] Move FPGA memory access enum definition (#1080)

Since there's a dedicated `// Enums` section in the `OCLUtil.h` header,
move the enum definition there.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

Fznamznon committed:
Add support of new IR pattern to transOCLBuiltinFromVariable (#1091)

* Add support of new IR pattern to transOCLBuiltinFromVariable

Some LLVM optimizations started generation of ascast -> gep -> load
sequence, this patch adds support for it.

Commit:
Fix a crash when alias metadata is attached to a call to lifetime inrinsic (#1059)

* Fix a crash when alias metadata is attached to a call to lifetime intrinsic

The correct translation of this case will be done later, once the spec
is updated.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

svenvh committed:
Add SPV_KHR_integer_dot_product support

Add support for mapping the SPV_KHR_integer_dot_product extension
operations to and from SPIR-V friendly IR.

Fznamznon committed:
Implement SPV_INTEL_debug_module extension (#1089)

* Implement SPV_INTEL_debug_module extension

Spec https://github.com/intel/llvm/pull/3976

svenvh committed:
Run spirv-val for more tests

svenvh committed:
[NFC] Remove unnecessary switch case for OpSelect

With SPIRVSelect inheriting from SPIRVInstTemplateBase, this can now
be handled through the switch default.

svenvh committed:
[NFC] Sync spirv.hpp and regenerate SPIRVNameMapEnum.h

Bring spirv.hpp in sync with ddf3230 ("Merge pull request #222 from
kpet/github-actions", 2021-06-30) from
github.com/KhronosGroup/SPIRV-Headers .

This brings the SPV_KHR_bit_instructions and SPV_INTEL_debug_module
extension enum values.  Drop the internal enum value of the latter.

Commit:
Implemented SPV_INTEL_runtime_aligned extension

This extension introduces a new function parameter attribute that can be
applied to pointers to indicate that the pointer was allocated with an
implementation-specific alignment.

Spec:
https://github.com/KhronosGroup/SPIRV-Registry/blob/c7079ba821f5ec70332ad6ec3a4962bfcd7f17ac/extensions/INTEL/SPV_INTEL_runtime_aligned.asciidoc

Signed-off-by: Leonid Pauzin <leonid.pauzin@intel.com>
Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

svenvh committed:
Report disallowed extension as error instead of assert

KornevNikita committed:
Fix using of deprecated CreateInBounds

Enable using of CreateInBounds with element type due to deprecate
GetElementPtrInst::CreateInBounds without element type.

KornevNikita committed:
Apply suggestion

KornevNikita committed:
Implement SPV_INTEL_arithmetic_fence (#1099)

Implement SPV_INTEL_arithmetic_fence

This extension creates support for a __fence keyword to prevent reordering and
reassociating operations.

Spec: https://github.com/intel/llvm/pull/3914

mlychkov committed:
Add support for SPV_INTEL_bfloat16_conversion extension

This extension provides instructions to convert single-precision 32-bit
floating-point value to bfloat16 format and vice versa.
It doesn't introduce bfloat16 type in SPIR-V, instead instructions below
use 16-bit integer type whose bit pattern represents (bitcasted from)
a bfloat16 value.

Spec: https://github.com/intel/llvm/blob/f587bbfb8a742bd55c80b12e01505ed085a0748f/sycl/doc/extensions/SPIRV/SPV_INTEL_bf16_convert.asciidoc

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

Commit:
Revert "[RFC] Improve code for IVDep backwards translation (#1084)"

Due to Windows-specific sporadic translation failures, revert until
root cause is identified.

This reverts commit 348d3cf97c121cab056331470cf761bad056c03c.

Commit:
Migrate arithmetic fence tokens to spirv_internal.hpp

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

svenvh committed:
Add cl_khr_extended_bit_ops support

This adds translation of the following OpenCL builtin functions to and
from SPIR-V operations:

 * bitfield_insert() <-> OpBitFieldInsert
 * bitfield_extract_signed() <-> OpBitFieldSExtract
 * bitfield_extract_unsigned() <-> OpBitFieldUExtract
 * bit_reverse() or llvm.bitreverse <-> OpBitReverse

OpBitReverse is translated into the llvm.bitreverse intrinsic, instead
of the bit_reverse function, which was already the translator's
behaviour before this change.

Co-authored-by: Stuart Brady <Stuart.Brady@arm.com>

svenvh committed:
Update for LLVM GEP API change

Update for LLVM commit b00cff56cfb1 ("Reapply [IR] Don't accept
nullptr as GEP element type", 2021-07-09).

svenvh committed:
Fix readonly attribute after new LLVM assertion

After LLVM commit 5d1464cbfe90 ("[Attributes] Make type attribute
handling more generic (NFCI)", 2021-07-12), the Attribute constructor
became stricter.  This revealed that the translator was creating
`readonly` attributes with a type attached, tripping the newly added
assertion in `Attribute::get`.

Commit:
Don't add aliasing MemAccess mask if metadata translation failed

Otherwise we might face an assert in case if empty metadata was generated by LLVM opt.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

Commit:
Add mapping for DWARF language codes to SourceLanguage (#1074)

When translating from LLVM IR to SPIR-V, the SourceLanguage is currently
set to the DWARF language value, which is wrong. This PR adds a mapping
from DWARF language codes to the correct value in the SourceLanguage enum.

Also comment out a test referring to C++ for OpenCL.
